rdr_service/offline/import_hpo_lite_pairing.py

```diff
@@ -0,0 +1,124 @@
+
+        new_obj = deepcopy(existing_obj)
+        new_obj.hpoId = hpo_id
+        new_obj.organizationId = org_id
+
+        self.participant_dao.update_with_session(session, new_obj)
+
+    def import_pairing_data(self, since: datetime = None):
+        """
+        :param since: DateTime to use as start of date range request. Will import all records created or modified
+            after the given date. Defaults to the start (midnight) of yesterday.
+        """
+        if since is None:
+            now_yesterday = datetime.now() - timedelta(days=1)
+            since = datetime(now_yesterday.year, now_yesterday.month, now_yesterday.day)
+
+        redcap = RedcapClient()
+        records = redcap.get_records(self.redcap_api_key, since)
+
+        with self.base_dao.session() as session:
+            for record in records:
+                try:
+                    if record.get('hpo_lite_pairing_complete') != '2':
+                        continue
+
+                    participant_id = from_client_participant_id(record.get('recordid'))
+
+                    org_name_mapping = config.getSettingJson(config.HPO_LITE_ORG_NAME_MAPPING)
+                    org_external_id = org_name_mapping.get(record.get('hpo_name'), None)
+                    org = self.org_dao.get_by_external_id(org_external_id)
+                    if not org:
+                        logging.error(f'Organization {org_external_id} not found for {participant_id}')
+                        continue
+
+                    upload_user_email = record.get('user_email')
+                    if upload_user_email and upload_user_email != '':
+                        upload_user = self.api_user_dao.load_or_init(USER_SYSTEM, upload_user_email)
+
+                    pair_date_str = record.get('paired_date')
+                    if pair_date_str:
+                        pair_date = parser.parse(pair_date_str)
+
+                    import_record = self._find_latest_import_record(participant_id, session)
+
+                    if import_record and import_record.orgId == org.organizationId:
+                        continue
+
+                    if not self._is_hpo_unset(session, participant_id):
+                        logging.error(f'{participant_id} is not exist or already paired with other HPO')
+                        continue
+
+                    # update participant pair information
+                    self._update_pairing_with_session(session, participant_id, org.hpoId, org.organizationId)
+
+                    # save the import record
+                    import_record = HpoLitePairingImportRecord(
+                        participantId=participant_id,
+                        pairedDate=pair_date,
+                        orgId=org.organizationId,
+                        created=CLOCK.now()
+                    )
+                    import_record.uploadingUser = upload_user
+
+                    self.base_dao.insert_with_session(session, import_record)
```

Comment:
What do you think of committing the session after using it for the record (right after the insert, before the `for` loop starts on the next record)?

That way, if the following records have any errors then previously successful pairings don't get rolled back. And it will also release the locks on the participant taken by the `_update_pairing_with_session` method's call to get it for update.

Reply:
Good point, done!